### PR TITLE
[DependencyInjection] Add `constructor` option to `#[Autoconfigure]`

### PR DESCRIPTION
**Pull request by alexandre-daubois**
| Q             | A
| ------------- | ---
| Branch?       | 6.3
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | _NA_
| License       | MIT
| Doc PR        | https://github.com/symfony/symfony-docs/pull/18070

Following [this discussion](https://twitter.com/nicolasgrekas/status/1634259754104025088?s=20) on Twitter, here is my try to use service factories with attributes. This PR adds the `constructor` option to the `Autoconfigure` attribute, and more generally, the `constructor` option on service definitions.

This allows to write services using a factory this way:

```php
#[Autoconfigure(constructor: 'create')]
class FactoryAttributeService
{
    public function __construct(private readonly string $bar)
    {
    }

    public function getBar(): string
    {
        return $this->bar;
    }

    public static function create(string $foo = 'foo'): static
    {
        return new self($foo);
    }
}
```

### CODE REVIEW

**Comment by GromNaN** [comment]
It would be very nice to have an even shorter syntax when the factory is a static function of the current class.

@weaverryan [proposed](https://twitter.com/weaverryan/status/1634266827105746944)
```php
#[Factory('create', ['$foo' => 'foo'])]
class FactoryAttributeService
{
```

The attribute could also be on the factory function (we will have to deal with duplicate).
```php
class FactoryAttributeService
{
    #[Factory(['$foo' => 'foo'])]
    public static function create(string $foo): static
```

**Comment by alexandre-daubois** [comment]
Thanks for the feedback @GromNaN! I implemented the first idea, but for the second one, I'm not sure.

I mean, this is indeed a great idea, but it seems it needs a lot of rework in `src/Symfony/Component/DependencyInjection/Compiler/RegisterAutoconfigureAttributesPass.php`. The pass loops through class attributes which makes it more complex to deal with the `#[Factory]` attribute on methods.

But maybe my approach is not the right one and there's something I don't see?

**Review by GromNaN — APPROVED** [review]
Awesome. Actually I have no idea how to implement the 2nd idea.

**Comment by stof** [comment]
Do we need this `$bind` argument ? Can't we rely on autowiring on the factory signature ?

**Comment by alexandre-daubois** [comment]
I don't know if it's a necessity, but being able to configure the specific bindings of the factory directly in the attribute seems to me a good idea, because it allows to have all the information "centralized in the same place" in some way. What do you think?

**Review by nicolas-grekas — CHANGES_REQUESTED** [review]
Thanks for starting this.

We didn't list "factory" in the possible options to autoconfigure because it just didn't make sense. The reason why it didn't make sense is that configuring a factory without telling the factory which class to instantiate doesn't make sense either.

It looks like we spotted the *only* situation where we do have a way to tell the factory which class it needs to instantiate : static constructors (the class is passed to the factory via `static::class`.)

If we want to support this pattern, I suggest this: `#[Autoconfigure(constructor: 'theMethod')]` (and I'm not sure we need a dedicated attribute for this.)

While adding support for this via attribute, we should also support it via other configuration means. Yaml/XML/PHP-DSL should support this "constructor" key for their "instance-of-conditionals" (which is what autoconfiguration works on).

About the attached patch I'd also like to have an end-to-end test case, one in PhpDumperTest and one in ContainerBuilderTest.

**Comment by donquixote** [comment]
I am not sure what is the goal here, I could see two:
- Discover services based on attributes, extending the `resource:` mechanism so it also works for factories.
- Provide additional information for services that are already being declared somewhere.

I am most interested in the first point.
For that, it would be most natural to have the attribute on a static method, and use the return type as the service id (unless a custom id is provided).
When doing this, I would name the attribute class `Service`, not `Factory`, because the fact it is a factory is already obvious.
A class can contain more than one factory, each returning a different service.

```php
class C {
  #[Service]
  static function createA(): A {..}
  #[Service]
  static function createB(): B {..}
}
```

To avoid duplicates, the attribute can have a custom id, or a suffix that will be appended to the class name.

```php
class C {
  #[Service]
  static function createA(): A {..}
  #[Service(id: 'a.other')]
  static function createOtherA(): A {..}
  #[Service]
  static function createB(): B {..}
  #[Service(suffix: '.other')]
  static function createOtherB(): B {..}
}
```

The `#[Service]` attribute could also be added on class level to explicitly register a class as a service.
I think right now this is not needed because `resource:` already registers all the classes. But perhaps there could be a different mode where this does not happen, and where only classes with `#[Service]` attribute are registered as services.

**Comment by alexandre-daubois** [comment]
Thank you for your feedback @nicolas-grekas! I completely reworked the PR. It now allows a `constructor` option on Autoconfigure. Also, this PR makes available `constructor` in PHP-DSL, YAML and XML.

When using `constructor`, `factory` cannot be used for obvious reasons.

@donquixote I think it would worth it to discuss this in an issue to gather feedback, as this PR's initial goal changed a bit 😄

File: src/Symfony/Component/DependencyInjection/Compiler/ResolveStaticConstructorPass.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+namespace Symfony\Component\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Definition;
+use Symfony\Component\DependencyInjection\Exception\RuntimeException;
+
+/**
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+class ResolveStaticConstructorPass extends AbstractRecursivePass
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ResolveStaticConstructorPass.php]: We don't need a new pass to handle this - and we don't need a new property on Definition: `setFactory([null, 'theMethod'])` is enough to hold the needed state.

Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Compiler/ResolveStaticConstructorPass.php]: I removed the pass and updated Defintion to remove the new property 👍

File: src/Symfony/Component/DependencyInjection/Definition.php [hunk 2]
```diff
@@ -123,6 +123,28 @@ public function getFactory(): string|array|null
...
     }
 
+    /**
+     * Sets a constructor.
+     *
+     * @param string|null $constructor the name of the public static function of the service to call to create itself
+     *
+     * @return $this
+     */
+    public function setConstructor(string|null $constructor): static
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Definition.php]: both methods should be removed

File: src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/StaticConstructorTrait.php [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\DependencyInjection\Loader\Configurator\Traits;
+
+trait StaticConstructorTrait
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/Configurator/Traits/StaticConstructorTrait.php]: ConstructorTrait

File: src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php [hunk 4]
```diff
@@ -314,6 +315,14 @@ private function parseDefinition(\DOMElement $service, string $file, Definition
             }
         }
 
+        if ($constructor = $this->getChildren($service, 'constructor')) {
+            if (null !== $definition->getFactory()) {
+                throw new LogicException(sprintf('The "%s" service cannot declare a factory as well as a constructor method.', $service->getAttribute('id')));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/XmlFileLoader.php]: ~method~

File: src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php [hunk 5]
```diff
@@ -501,6 +505,14 @@ private function parseDefinition(string $id, array|string|null $service, string
             $definition->setFactory($this->parseCallable($service['factory'], 'factory', $id, $file));
         }
 
+        if (isset($service['constructor'])) {
+            if (null !== $definition->getFactory()) {
+                throw new LogicException(sprintf('The "%s" service cannot declare a factory as well as a constructor method.', $id));
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/YamlFileLoader.php]: ~method~

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd [hunk 6]
```diff
@@ -147,6 +151,7 @@
       <xsd:element name="argument" type="argument" minOccurs="0" maxOccurs="unbounded" />
       <xsd:element name="configurator" type="callable" minOccurs="0" maxOccurs="1" />
       <xsd:element name="factory" type="factory" minOccurs="0" maxOccurs="1" />
+      <xsd:element name="constructor" type="constructor" minOccurs="0" maxOccurs="1" />
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: same line is missing in `<xsd:complexType name="instanceof">` and in prototype

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveChildDefinitionsPassTest.php [hunk 7]
```diff
@@ -438,4 +438,21 @@ public function testProcessCopiesSyntheticStatus()
         $def = $container->getDefinition('child');
         $this->assertTrue($def->isSynthetic());
     }
+
+    public function testProcessCopiesStaticConstructor()
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/ResolveChildDefinitionsPassTest.php]: this test case is not needed

File: src/Symfony/Component/DependencyInjection/Loader/Configurator/ServiceConfigurator.php [hunk 8]
```diff
@@ -36,6 +36,7 @@ class ServiceConfigurator extends AbstractServiceConfigurator
     use Traits\PropertyTrait;
     use Traits\PublicTrait;
     use Traits\ShareTrait;
+    use Traits\StaticConstructorTrait;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/Configurator/ServiceConfigurator.php]: there is also InstanceofConfigurator

File: src/Symfony/Component/DependencyInjection/CHANGELOG.md [hunk 9]
```diff
@@ -18,6 +18,7 @@ CHANGELOG
  * Add support for autowiring services as closures using `#[AutowireCallable]` or `#[AutowireServiceClosure]`
  * Deprecate `#[MapDecorated]`, use `#[AutowireDecorated]` instead
  * Deprecate the `@required` annotation, use the `Symfony\Contracts\Service\Attribute\Required` attribute instead
+ * Add `constructor` option to `#[Autoconfigure]`
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/CHANGELOG.md]: not only to the attribute now

File: src/Symfony/Component/DependencyInjection/Tests/Compiler/RegisterAutoconfigureAttributesPassTest.php [hunk 10]
```diff
@@ -47,6 +48,7 @@ public function testProcess()
             ->addTag('another_tag', ['attr' => 234])
             ->addMethodCall('setBar', [2, 3])
             ->setBindings(['$bar' => $argument])
+            ->setConstructor('create')
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Compiler/RegisterAutoconfigureAttributesPassTest.php]: why this change?

Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Tests/Compiler/RegisterAutoconfigureAttributesPassTest.php]: This change (now updated with `setFactory`) ensures that the `constructor` option in `#[Autoconfigure]` of `AutoconfigureAttributed` is taken into account by the pass.

Also, I updated everything else according to your feedback 🙏

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd [hunk 11]
```diff
@@ -126,6 +126,10 @@
     <xsd:attribute name="expression" type="xsd:string" />
   </xsd:complexType>
 
+  <xsd:complexType name="constructor">
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: instead of a node, shouldn't we use an attribute? `<service constructor="create" />`

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/constructor.xml
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/constructor.xml:1]: file should be named static_constructor
can you also add a test covering the php-dsl a bit

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services_with_constructor.yml
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services_with_constructor.yml:1]: what is this file for?

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd (hunk 11 above)
Comment (stof) [reply src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: this should indeed be an attribute

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services_with_constructor.yml
Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/yaml/services_with_constructor.yml:1]: Artifact, removed :+1:

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/constructor.xml
Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/xml/constructor.xml:1]: Added! I'm currently fixing the last test :+1:

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd (hunk 11 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: Updated!

Comment (nicolas-grekas) [reply src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: You forgot to remove this "complex type" ;)

Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: Arf sorry 🤦 Removed as well and tests are fixed!

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd [hunk 12]
```diff
@@ -184,6 +185,7 @@
     <xsd:attribute name="lazy" type="xsd:string" />
     <xsd:attribute name="autowire" type="boolean" />
     <xsd:attribute name="autoconfigure" type="boolean" />
+    <xsd:attribute name="constructor" type="boolean" />
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: xsd:string

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd [hunk 13]
```diff
@@ -208,6 +210,7 @@
     <xsd:attribute name="parent" type="xsd:string" />
     <xsd:attribute name="autowire" type="boolean" />
     <xsd:attribute name="autoconfigure" type="boolean" />
+    <xsd:attribute name="constructor" type="boolean" />
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: xsd:string

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/config/inline_static_constructor.expected.yml [hunk 14]
```diff
@@ -0,0 +1,10 @@
+
+services:
+    service_container:
+        class: Symfony\Component\DependencyInjection\ContainerInterface
+        public: true
+        synthetic: true
+    foo:
+        class: Symfony\Component\DependencyInjection\Tests\Fixtures\Prototype\StaticConstructor\PrototypeStaticConstructorAsArgument
+        public: true
+        arguments: [!service { class: Symfony\Component\DependencyInjection\Tests\Fixtures\Prototype\StaticConstructor\PrototypeStaticConstructor, factory: [Symfony\Component\DependencyInjection\Tests\Fixtures\Prototype\StaticConstructor\PrototypeStaticConstructor, create] }]
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Tests/Fixtures/config/inline_static_constructor.expected.yml]: this makes me wonder if we couldn't add a rule to dump a constructor here (see YamlDumper, same in XmlDumper)

File: src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd (hunk 12 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Loader/schema/dic/services/services-1.0.xsd]: Of course, I've been too fast. That's updated, and the other one as well!

File: src/Symfony/Component/DependencyInjection/Tests/Fixtures/config/inline_static_constructor.expected.yml (hunk 14 above)
Comment (alexandre-daubois) [reply src/Symfony/Component/DependencyInjection/Tests/Fixtures/config/inline_static_constructor.expected.yml]: That's updated! 👌

**Review by nicolas-grekas — APPROVED** [review]
LGTM thanks (but a rebase is needed.)

**Comment by alexandre-daubois** [comment]
That's rebased :+1: !

**Comment by fabpot** [comment]
Thank you @alexandre-daubois.

**Comment by kaznovac** [comment]
I'm late to the game so feel free to ignore this comment.

The `constructor` property name is confusing as it points to the existing (and in this case wrong) OOP concept.
It also rises the question "How the services are instantiated if this property is not set - without the constructor?!"

`staticFactoryMethod` is the name of this instantiation concept.
I agree it's too mouthful for the property name but it points a developer in the correct direction.
